log.go

```diff
@@ -29,6 +29,35 @@ type raftLog struct {
 	// they will be saved into storage.
 	unstable unstable
 
+	// leaderTerm is a term of the leader with whom our log is "consistent". The
+	// log is guaranteed to be a prefix of this term's leader log.
+	//
+	// The leaderTerm can be safely updated to `t` if:
+	//	1. the last entry in the log has term `t`, or, more generally,
```

Comment:
Only (2) is necessary.

On a server restart though, we forget who did the last append. (1) gives the best guess, and allows recovering to up-to-date state for idle raft groups that haven't seen recent appends except the current leader's empty entry. For non-idle groups, recovering to `lastTerm` gives no value, and is equivalent to setting `leaderTerm = 0`.

It would be more ideal if this field was stored in some `HardState.LeaderTerm` - then we would always recover to up-to-date state. Note that we can't reuse `HardState.Term` because `HardState.Term` can be > `leaderTerm` (for the same reason why `r.Term` can be > `leaderTerm`).

To bring analogy with Paxos, the`raftLog` is an acceptor, and `leaderTerm` is the id of the highest accepted proposal. The election term is sort of orthogonal to this - the leader term can jump in front of the accepted proposal term.

Reply:
I should note the invariant here:

```
	RawNode.raft.Term >= raftLog.leaderTerm >= raftLog.lastTerm()
```

(1) initializes `leaderTerm` to `lastTerm()`, which is safe because `raft.Term >= lastTerm()`
(2) maintains it

Reply:
> allows recovering to up-to-date state for idle raft groups that haven't seen recent appends except the current leader's empty entry. For non-idle groups, or followers whose log is significantly behind the leader's and doesn't end with entries at its term, recovering to raftLog.lastTerm() gives no value, and is equivalent to setting leaderTerm = 0. To recover, these will then have to "wait" for the next append message from this leader.

This is what I was hoping to clarify. Initializing to `lastTerm` is an opportunistic way to allow a restarted follower in idle raft groups to immediately advance its commit index on startup without needing to first accept a MsgApp.

Is it anything more than that? For an idle raft group, a follower with an up-to-date log but a lagging commit index may restart and never receive any new entries. If we didn't have (1) and we started discarding commit indexes in heartbeats with terms > raftLog.leaderTerm, would the commit index on the follower get stuck?

Reply:
That's a good point. I think we indeed need (1). Also, (1) plays nicely with the invariant that I put above.

Reply:
> If we ever want to bring this to the next level: `MsgApp` messages should not be rejected based on `r.Term / HardState.Term`. For correctness, it is only necessary to reject `MsgApp` if the message term is < `raftLog.leaderTerm / HardState.LeaderTerm`. I think this would reduce some unnecessary rejects during the leader election flux time.

Does this mean MsgApp from old leader will be accepted by voters that voted to a new leader but yet received any log from the new leader? In that case, there may be committed entry in old leader but not in the new leader.

Reply:
@joshuazh-x That's a good point. We don't want a quorum of nodes to accept entries unless we are sure these entries are consistent with the new leader's log.

So it's safest to accept only `MsgApp.Term >= raft.Term`. But we could sometimes accept `raftLog.leaderTerm <= MsgApp.Term < raft.Term` **if**:

- The `MsgApp` contains the `(index, term)` entry for which we voted when electing the `raft.Term` leader.
- We would cut this append at the aforementioned `index`.
- This guarantees that this append is consistent with the new leader's log.

It's unclear if such an optimization would give any value (like reduce commit latency in some cases), so I will leave it as an exercise for later :) Looks like a complication.